File: modulo-01-fundamentos-arquitectura/codigo/src/main/java/pe/joedayz/microservicios/modulo01/ddd/order/OrderLine.java
package pe.joedayz.microservicios.modulo01.ddd.order;

import pe.joedayz.microservicios.modulo01.ddd.shared.Money;
import pe.joedayz.microservicios.modulo01.ddd.shared.Quantity;
import pe.joedayz.microservicios.modulo01.ddd.shared.Sku;

/**
 * Linea de un pedido. Vive DENTRO del agregado {@link Order} y no se manipula por
 * fuera: se crea a traves de {@code order.addLine(...)}, que valida los invariantes.
 *
 * <p>Es inmutable; representa "N unidades del SKU X a un precio unitario".
 */
public record OrderLine(Sku sku, Quantity quantity, Money unitPrice) {

    public Money subtotal() {
        return unitPrice.times(quantity.value());
    }
}

File: modulo-01-fundamentos-arquitectura/codigo-dotnet/src/Modulo01.Fundamentos/Ddd/Orders/OrderLine.cs
using JoeDayz.Microservicios.Modulo01.Ddd.Shared;

namespace JoeDayz.Microservicios.Modulo01.Ddd.Orders;

/// <summary>
/// Entidad interna del agregado: nunca se modifica desde fuera, solo a traves de <see cref="Order"/>.
/// </summary>
public sealed record OrderLine(Sku Sku, Quantity Quantity, Money UnitPrice)
{
    public Money Subtotal() => UnitPrice.Times(Quantity.Value);
}
